untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+$and\;fls\;tru$
+$= (\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. b)\;(\lambda c\; d. c)$
+$\to (\lambda y. (\lambda a\; b. b)\;y\;fls)\;(\lambda c\; d. c)$
+$\to (\lambda a\; b. b)\;(\lambda c\; d. c)\;fls$
+$\to (\lambda b. b)\;fls$
+$\to fls.$
+
+That also works.
+The remaining two cases are left as an exercise for the reader, but it should be rather obvious that the combinator works as intended.
+Indeed, since $tru$ essentially chooses its first argument and $fls$ its second, $x\;y\;fls$ essentially means "if $x$ is $tru$, the result is $y$, otherwise $fls$".
+
+Defining other boolean combinators is reasonably straightforward by analogy.
+
+### Church pairs
+
+Since we have booleans now, we can also encode pairs (and by extension, tuples as trees or right combs) as functions, returning one value when given $tru$ as an argument, and another when given $fls$.
+
+First of all, let us define a pair constructor combinator:
+
+$pair = \lambda f\;s\;b. b\;f\;s.$
+
+It looks suspiciously similar to our $ifThenElse$ combinator, the only difference being that condition is now the last argument.
+
+Now we can define two combinators for deconstructing a pair:
+
+$fst = \lambda p. p\;tru,$\
+$snd = \lambda p. p\;fls.$
+
+It's reasonably straightforward to show that these definitions work as intended.
+Consider, for instance,
+
+$fst\;(pair\;x\;y)$
+$= (\lambda p. p\;tru)\;((\lambda f\;s\;b. b\;f\;s)\;x\;y)$
+$\Rightarrow (\lambda p. p\;tru)\;(\lambda b. b\;x\;y)$
+$\to (\lambda b. b\;x\;y)\;tru$
+$\to tru\;x\;y$
+$= (\lambda a\; b. a)\;x\;y$
+$\Rightarrow x.$
+
+Here, double arrow $\Rightarrow$ signifies more than one step of beta-reduction.
+
+### Church naturals
+
+Since we encoded both pairs and booleans, we already can encode naturals simply as sequences of booleans, that is, as binary numbers.
+In practice, this turns out to be cumbersome, however, so instead, we'll define naturals similar to the Peano construction.
+Essentially, we'll be choosing an encoding for zero and then encoding other numbers as some successor function applied to zero.
+
+Church used the following definitions:
+
+$c_0 = \lambda s\;z. z\;$\
+$c_1 = \lambda s\;z. s\;z,$\
+$c_2 = \lambda s\;z. s\;(s\;z),$\
+etc.
+
+The beautiful thing about this encoding is that it naturally encodes the concept of "counting".
+Each Church natural takes a "next" combinator $s$ and a "starting value" $z,$ and then "counts" the corresponding "nexts" from $z.$
+
+Now we can define a successor function, taking a natural and returning the next natural.
+Since we defined naturals as functions of two arguments, we expect the successor function to be a function of three arguments (i.e. three nested abstractions).
+The body should apply the "next" argument to the natural one more time:
+
+$succ = \lambda n\; s\; z. s\;(n\;s\;z).$
+
+It might be easier to see what's going on if we explicitly note the outer abstraction:
+
+$succ = \lambda n. \lambda s\; z. s\;(n\;s\;z).$
```

Comment:
Some of these are defined with $...$ and some with $$...$$. 

Since they render differently, I think it would be better to just use $$....$$ for all the math blocks that are not inside text.

Reply:
Here, we can, yes.